src/deploy/functions/validate.ts

```diff
@@ -0,0 +1,71 @@
+"use strict";
+
+import * as FirebaseError from "../../error";
+import * as _ from "lodash";
+import * as path from "path";
+import * as clc from "cli-color";
+import * as logger from "../../logger";
+import * as projectPath from "../../projectPath";
+import * as fsutils from "../../fsutils";
+
+// have to require this because no @types/cjson available
+// tslint:disable-next-line
+const cjson = require("cjson");
+
+// Check that functions directory exists
+export function functionsDirectoryExists(cwd: string, sourceDirName: string): void {
+  if (!fsutils.dirExistsSync(projectPath.resolveProjectPath(cwd, sourceDirName))) {
+    const msg =
+      `could not deploy functions because the ${clc.bold('"' + sourceDirName + '"')} ` +
+      `directory was not found. Please create it or specify a different source directory in firebase.json`;
+    throw new FirebaseError(msg);
+  }
+}
+
+// Validate function names only contain lower case letters, numbers, and dashes
+export function functionNamesAreValid(functionNames: string[]): void {
+  const validFunctionNameRegex = /^[a-z][a-zA-Z0-9_-]{1,62}$/i;
+  const invalidNames = _.reject(
+    _.keys(functionNames),
+    (name: string): boolean => {
+      return _.startsWith(name, ".") || validFunctionNameRegex.test(name);
+    }
+  );
+  if (!_.isEmpty(invalidNames)) {
+    const msg = `${invalidNames.join(
+      ", "
+    )} function name(s) must be a valid subdomain (lowercase letters, numbers and dashes)`;
+    throw new FirebaseError(msg);
+  }
+}
+
+// Validate contents of package.json - main file present and engines specified
+export function packageJsonIsValid(
+  sourceDirName: string,
+  sourceDir: any,
+  projectDir: string
+): void {
+  const packageJsonFile = path.join(sourceDir, "package.json");
+  if (fsutils.fileExistsSync(packageJsonFile)) {
+    try {
+      const data = cjson.load(packageJsonFile);
+      logger.debug("> [functions] package.json contents:", JSON.stringify(data, null, 2));
+      const indexJsFile = path.join(sourceDir, data.main || "index.js");
+      if (!fsutils.fileExistsSync(indexJsFile)) {
+        const msg = `${path.relative(
+          projectDir,
+          indexJsFile
+        )} does not exist, can't deploy Firebase Functions`;
+        throw new FirebaseError(msg);
+      }
+    } catch (e) {
+      const msg = `There was an error reading ${sourceDirName}${path.sep}package.json:\n\n ${
+        e.message
+      }`;
+      throw new FirebaseError(msg);
+    }
+  } else if (!fsutils.fileExistsSync(path.join(sourceDir, "function.js"))) {
```

Comment:
*<dusts off parts of my brain I haven't used for a while>*

I believe this had something to do with Cloud Functions defaulting to `function.js` instead of `index.js` and we were trying to make our behavior consistent with their behavior...I super don't remember this well but it's something to the effect of "if there's no package.json but there is a function.js, it's still deployable".

Reply:
Hmm, sounds like legacy behavior - I think best thing would be to leave a comment explaining this and leave the code as is. Thoughts?

Reply:
If that's true, leaving it should be fine, I suppose. If Cloud Functions doesn't do this behavior any more, we should eventually come back and remove it

Reply:
I'm actually comfortable ditching this behavior now. The only way this would possibly work is if the developer had manually constructed a `__trigger` annotation on an object exported from `function.js` with no other dependencies. That's not a case we need to support.

That also makes this cleaner overall -- you can move the `No package found` throw to the top and ditch the large conditional altogether.

Reply:
Sounds great! Removed.